finite_difference/src/subdomain_mod.f90

```diff
@@ -0,0 +1,302 @@
+  !! domain decomposition
+  type :: decomposition_type
+     !> Dimensions of the grid of sub-domains
+     integer :: nx, ny
+     !> Number of sub-domains (=nx*ny)
+     integer :: ndomains
+     !> Max dimensions of any sub-domain
+     integer :: max_width, max_height
+     !> Array of the sub-domain definitions
+     type(subdomain_type), allocatable :: subdomains(:)
+  end type decomposition_type
+
+contains
+
+  function decompose(domainx, domainy,             &
+                     ndomains, ndomainx, ndomainy, &
+                     halo_width) result(decomp)
+    !> Decompose a domain consisting of domainx x domainy points
+    !! into a 2D grid.
+    !! Returns an array of tiles describing each of the subdomains.    
+    use parallel_mod, only: get_num_ranks, parallel_abort
+    implicit none
+    !> The decomposition that this function will return
+    type(decomposition_type), target :: decomp
+    !> Dimensions of the domain to decompose
+    integer, intent(in) :: domainx, domainy
+    !> No. of domains to decompose into. If not supplied will use the
+    !! number of MPI ranks (or 1 if serial).
+    integer, intent(in), optional :: ndomains
+    !> Optional specification of the dimensions of the tiling grid
+    integer, intent(in), optional :: ndomainx, ndomainy
+    !> Width of halo to allow for when constructing sub-domains.
+    !! Default value is 1. Must be > 0 for a multi-processor run.
+    integer, intent(in), optional :: halo_width
+    ! Locals
+    integer :: ival, jval ! For tile extent calculation
+    integer :: internal_width, internal_height
+    integer :: ierr, nwidth
+    integer :: ji,jj, ith
+    integer :: jover, junder, height
+    integer :: iover, iunder, width
+    ! For doing stats on tile sizes
+    integer :: nvects, nvects_sum, nvects_min, nvects_max 
+    logical, parameter :: print_tiles = .TRUE.
+    ! Whether to automatically compute the dimensions of the tiling grid
+    logical :: auto_tile = .TRUE.
+    integer :: xlen, ylen
+    integer :: ntilex, ntiley
+    ! Local var to hold number of sub-domains
+    integer :: ndom = 1
+    ! Local var to hold halo width
+    integer :: hwidth = 1
+    ! Pointer to the current subdomain being defined
+    type(subdomain_type), pointer :: subdomain
+
+    ! Deal with optional arguments to this routine
+    if(.not. present(ndomains))then
+       if(.not. present(ndomainx) .and. .not. present(ndomainy))then
+          ! Automatically use the number of MPI processes
+          ndom = get_num_ranks()
+          auto_tile = .TRUE.
+       else if(present(ndomainx) .and. present(ndomainy))then
+          ndom = ndomainx * ndomainy
+          auto_tile = .FALSE.
+       else
+          call parallel_abort('decompose: invalid arguments supplied')
+       end if
+    else
+       ndom = ndomains
+       auto_tile = .TRUE.
+    end if
+
+    if(present(halo_width))then
+       if(halo_width < 1 .and. get_num_ranks() > 1)then
```

Comment:
I would imagine that there are some use cases for distributed fields without halo - e.g. if you just compute an output field from other fields (likely with halo). So it might be too restrictive to enforce those fields to have a halo.

Reply:
Oh yes, good point. However, this routine computes a general decomposition, independent of any particular field. This might be influenced by some of your points below though so I'll hold off for now.